Allow for cooperative release of funds

patrini32 opened the pull request:
Should close #693. Introduces new logic for the BtcPunished state.

patrini32 commented:
Oh, latest commit broke the code.

patrini32 commented on swap/src/protocol/bob/state.rs:
I added this function to State3 because it's possible that Bob's database is in incorrect state.

binarybaron commented:
@patrini32 Can you give @einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.

patrini32 commented:
> @patrini32 Can you give @Einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.

Sure.

binarybaron commented:
> > @patrini32 Can you give @Einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.
> 
> Sure.

thanks :)

patrini32 commented:
Sent invite

Einliterflasche commented:
thanks! I think using `let else` really pays off here

patrini32 commented:
> thanks! I think using `let else` really pays off here

Thank you for refactoring it.

binarybaron commented on swap/src/api/request.rs:
```diff
@@ -810,6 +816,159 @@ impl Request {
+                )
+                .await?;
+                tokio::spawn(event_loop.run().in_current_span());
+                tracing::info!("Asking Alice to reveal xmr key to us.");
```
nitpick: Remove the dot at the end

patrini32 replied:
Okay

binarybaron commented on swap/src/api/request.rs:
```diff
@@ -810,6 +816,159 @@ impl Request {
+                    .await;
+                match response {
+                    Ok(response) => {
+                        tracing::info!("Alice revealed xmr key to us, redeeming xmr...");
```
same here xmr -> XMR

binarybaron reviewed (changes requested):
This looks pretty good! I commented two small nitpicks. And two things we need to get done before we merge this:
1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished

patrini32 commented:
> This looks pretty good! I commented two small nitpicks. And two things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished

I implemented 2 initially, but it required changes to database.

patrini32 commented:
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

I implemented 2, but it requires changes to multiple BobStates, because state3 doesn't contain monero_wallet_restore_blockheight, so we can't just change State6.

patrini32 commented:
Hmm, actually, what if pass monero_wallet_restore_blockheight to cancel?

patrini32 commented:
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

I implemented 2 and added new test for it.
@binarybaron

patrini32 commented:
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

Implemented all of these, can you review, please? @binarybaron

patrini32 commented on migrate/src/main.rs:
```diff
@@ -0,0 +1,132 @@
+    let args = Args::from_args();
+    let path_str = &args.file_name.display();
+    let pool = SqlitePool::connect(&format!("sqlite:{}", path_str)).await?;
+    println!("Copying {path_str} to {path_str}.original...");
```
I don't think that we need tracing there.

patrini32 commented:
@binarybaron, is it okay to use sql migrations?

patrini32 commented:
> > @binarybaron, is it okay to use sql migrations?
> 
> Yes that's fine. I'll have to review the SQL in depth. Can you add comments to the SQL query?

Sure

patrini32 commented:
> > @binarybaron, is it okay to use sql migrations?
> 
> Yes that's fine. I'll have to review the SQL in depth. Can you add comments to the SQL query?

Should I add comments to all SQL queries or just the first two? (Others are copies of these).

patrini32 commented on swap/src/protocol/bob/swap.rs:
```diff
@@ -39,10 +37,12 @@ pub async fn run_until(
         swap.db
             .insert_latest_state(swap.id, current_state.clone().into())
             .await?;
+        if matches!(current_state, BobState::BtcPunished { .. }) {
```
Is it okay to add this to run_until or should this function be more general?

binarybaron commented:
@delta1 Do you have time to give this a review?

patrini32 commented:
Oh, sorry, @binarybaron, can you rerun the CI, I fixed tests.

patrini32 commented:
> Oh, sorry, @binarybaron, can you rerun the CI? I fixed tests.
Uh, that's weird. I ran this test several times and it didn't fail.

patrini32 commented:
@binarybaron, can you run CI again, I think there's some nondeterminism there.

binarybaron commented:
Can you also add some more general comments to the SQL migration file? The ones you added are good but are very specific. They are quite hard to understand without having as much context.

For example: "This query adds the new state3 attribute that has been added in PR..."

patrini32 commented:
> Can you also add some more general comments to the SQL migration file? The ones you added are good but are very specific. They are quite hard to understand without having as much context.
> 
> For example: "This query adds the new state3 attribute that has been added in PR..."

Done.

patrini32 commented:
Are there any other changes I should make?

patrini32 commented:
> After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> 
> ```
> ./target/debug/swap --testnet history
> Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> Error: missing field `v` at line 1 column 3327
> ```

Hmm, that's odd, it worked when I tested it.

binarybaron commented:
> > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > ```
> > ./target/debug/swap --testnet history
> > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > Error: missing field `v` at line 1 column 3327
> > ```
> 
> Hmm, that's odd, it worked when I tested it.

I'll investigate. Are you added the new v attribute to every state6 attribute in every state?

patrini32 commented:
> > > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > > ```
> > > ./target/debug/swap --testnet history
> > > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > > Error: missing field `v` at line 1 column 3327
> > > ```
> > 
> > 
> > Hmm, that's odd, it worked when I tested it.
> 
> I'll investigate. Are you added the new v attribute to every state6 attribute in every state?
Yes

patrini32 commented:
> > > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > > ```
> > > ./target/debug/swap --testnet history
> > > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > > Error: missing field `v` at line 1 column 3327
> > > ```
> > 
> > 
> > Hmm, that's odd, it worked when I tested it.
> 
> I'll investigate. Are you sure added the new v attribute to every state6 attribute in every state?

I tested it now. It works.

binarybaron commented:
Can you use an enum for the errors instead of a string?

patrini32 commented:
> Can you use an enum for the errors instead of a string?

Do you want to parse errors on the cli side?

binarybaron commented:
> > Can you use an enum for the errors instead of a string?
> 
> Do you want to parse errors on the cli side?

You can use `thiserror`

For example see how it is being used in `swap/src/network/swap_setup/alice.rs`

```rust
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("ASB is running in resume-only mode")]
    ResumeOnlyMode,
    #[error("Amount {buy} below minimum {min}")]
    AmountBelowMinimum {
        min: bitcoin::Amount,
        buy: bitcoin::Amount,
    },
    #[error("Amount {buy} above maximum {max}")]
    AmountAboveMaximum {
        max: bitcoin::Amount,
        buy: bitcoin::Amount,
    },
    #[error("Unlocked balance ({balance}) too low to fulfill swapping {buy}")]
    BalanceTooLow {
        balance: monero_rpc::wallet::GetBalance,
        buy: bitcoin::Amount,
    },
    #[error("Failed to fetch latest rate")]
    LatestRateFetchFailed(#[source] Box<dyn std::error::Error + Send + Sync + 'static>),
    #[error("Failed to calculate quote")]
    SellQuoteCalculationFailed(#[source] anyhow::Error),
    #[error("Blockchain networks did not match, we are on {asb:?}, but request from {cli:?}")]
    BlockchainNetworkMismatch {
        cli: BlockchainNetwork,
        asb: BlockchainNetwork,
    },
}
```

binarybaron commented on swap/migrations/20240615140942_btcpunished_update.sql:
```diff
@@ -0,0 +1,128 @@
+                        WHERE
+                            states.swap_id = swap_states.swap_id
+                            AND json_extract(
+                                states.state, '$.Bob.BtcCancelled'
```
We can enter the `BtcPunished` state without ever entering the `BtcCancelled` state. Can you extract this data from somewhere else?

binarybaron reviewed (changes requested):
sql query

binarybaron commented:
@Einliterflasche @delta1 Can you give this another glance before I merge this?

delta1 commented:
@binarybaron i will review tomorrow

delta1 reviewed (approved):
LGTM
